File: radiometry.py
#!/usr/bin/env python3
"""Temperature conversion for the Magnity camera — a faithful reimplementation of the
camera's built-in radiometric chain, reverse-engineered from libcoresdk.so.

DECODED CHAIN (the firmware's `sub_3f970` radiometric core; see PROTOCOL.md):

    v      = raw*(k+1) + b                       # k = ctx[0x50], b = ctx[0x54]  (floats)
    target = v << (7 - shift)                    # shift = ctx[0x775c] (small int)
    idx    = binary_search(radLUT, target)       # radLUT = 646-entry Planck radiance LUT
    U      = idx*4096 - 150000                    # temperature grid, in milli-Kelvin
             + ( slopeLUT[idx] * (target-radLUT[idx]) ) >> 12      # linear interp
    °C     = U/1000 - 273.15

What is FIXED in the hardware vs. what is per-camera:
  * `radLUT` (the Planck radiance curve) and `slopeLUT` are the FIXED part. The firmware
    keeps them in .bss (radLUT @0x29d700, slopeLUT @0x29e11c) and fills them at runtime
    from one of 8 static Planck curves baked into .rodata -> extracted to
    `recon/planck_luts.npy`. We reconstruct slopeLUT exactly as the firmware does:
    `slopeLUT[i] = 2**24 // (radLUT[i+1]-radLUT[i])`  (so (slope*Δ)>>12 == one idx step).
  * `k`, `b` (and the `shift` scaling) are the PER-CAMERA part. The firmware does NOT
    store them as constants — `sub_3d280` recomputes them every frame from the factory
    NUC blocks (per sensor-temperature, stride 0x21c), the live sensor temperature
    (ctx[0x80]) and the live shutter/FFC reference (`sub_3d9f4`). The NUC blocks are
    expanded from `mag_cali.bin` by a parser that lives in an external library behind
    C++ stream layers (`sub_34832` -> PLT import), so they can't be extracted offline.

CONSEQUENCE: we reproduce the FIXED radiometric curve exactly, and recover the per-camera
linear term (a, b) — which folds k, b and the `<<(7-shift)` scaling into
`radiance = a*raw + b` — from a few (raw, known_°C) reference points via `calibrate()`.
Because the curve is the camera's true Planck LUT (not a guessed shape), 2 references are
enough for accurate absolute temperature over a wide range; the firmware itself only
differs by re-deriving (a,b) live as the sensor drifts (which is what re-FFC + re-cal do).
"""
import os, numpy as np

_HERE = os.path.dirname(os.path.abspath(__file__))
LUT_PATH = os.path.join(_HERE, "recon", "planck_luts.npy")

U_OFFSET = -150000          # U = idx*4096 + U_OFFSET
U_STEP = 4096
KELVIN = 273.15
U_PER_KELVIN = 1000.0       # U is in milli-Kelvin (verified: idx109 -> 23.31 °C)


def celsius_to_U(tc):
    return (tc + KELVIN) * U_PER_KELVIN

def U_to_celsius(u):
    return u / U_PER_KELVIN - KELVIN


class Radiometry:
    def __init__(self, lut_path=LUT_PATH):
        self.luts = np.load(lut_path).astype(np.int64)     # (8, 646) int radiance curves
        self.n_lut, self.n_idx = self.luts.shape
        # firmware-exact reciprocal-slope tables: slopeLUT[i] = 2**24 // (L[i+1]-L[i])
        diffs = np.diff(self.luts, axis=1)                 # (8, 645)
        self.slopes = (1 << 24) // np.maximum(diffs, 1)
        self.a = 1.0
        self.b = 0.0
        self.lut_idx = self.n_lut // 2
        self.calibrated = False

    # radiance at a given U (temperature-grid value), via forward interpolation
    def _radiance_at_U(self, lut, u):
        pos = (np.asarray(u, float) - U_OFFSET) / U_STEP
        pos = np.clip(pos, 0, self.n_idx - 1.0001)
        i = pos.astype(int)
        frac = pos - i
        return lut[i] * (1 - frac) + lut[i + 1] * frac

    def _U_from_radiance(self, li, rad):
        """Inverse mapping radiance(target) -> U, BIT-FAITHFUL to sub_3f970's simple path:
        binary-search for idx in radLUT `li`, then U = idx*4096 - 150000
        + (slopeLUT[idx]*Δ)>>12.  (Matches the float interp to <0.01 °C.)"""
        lut = self.luts[li]; slope = self.slopes[li]
        ti = np.clip(np.floor(np.asarray(rad, float)).astype(np.int64), lut[0], lut[-1])
        idx = np.clip(np.searchsorted(lut, ti, side='right') - 1, 0, self.n_idx - 2)
        delta = ti - lut[idx]                               # target - radLUT[idx]  (>=0)
        u = idx.astype(np.int64) * U_STEP + U_OFFSET + ((slope[idx] * delta) >> 12)
        return u.astype(float)

    def raw_to_celsius(self, raw):
        """raw counts -> °C, using the calibrated linear term and the camera Planck LUT.
        `a*raw+b` folds the firmware's k,b and the <<(7-shift) scaling into one line."""
        rad = self.a * np.asarray(raw, float) + self.b
        u = self._U_from_radiance(self.lut_idx, rad)
        return U_to_celsius(u)

    def frame_to_celsius(self, raw_frame):
        """Vectorised whole-frame conversion (HxW raw -> HxW °C) for ROI/point tools."""
        f = np.asarray(raw_frame, float)
        return self.raw_to_celsius(f.ravel()).reshape(f.shape)

    def calibrate(self, points):
        """points: list of (raw_value, known_celsius). Tries every built-in LUT, fits a
        linear raw->radiance (a,b) by least squares so that LUT_inverse reproduces the
        known temps, and keeps the LUT with the smallest °C residual.
        Returns (rms_celsius, n_points)."""
        pts = [(float(r), float(t)) for r, t in points]
        if len(pts) < 2:
            raise ValueError("need >= 2 reference points")
        raws = np.array([p[0] for p in pts])
        temps = np.array([p[1] for p in pts])
        best = None
        for li in range(self.n_lut):
            lut = self.luts[li]
            # target radiance for each known temp (forward curve)
            target_rad = self._radiance_at_U(lut, celsius_to_U(temps))
            # fit radiance = a*raw + b  (least squares)
            A = np.vstack([raws, np.ones_like(raws)]).T
            (a, b), *_ = np.linalg.lstsq(A, target_rad, rcond=None)
            # residual measured back in °C
            self.a, self.b, self.lut_idx = a, b, li
            pred = self.raw_to_celsius(raws)
            rms = float(np.sqrt(np.mean((pred - temps) ** 2)))
            if best is None or rms < best[0]:
                best = (rms, li, a, b)
        rms, li, a, b = best
        self.a, self.b, self.lut_idx, self.calibrated = a, b, li, True
        return rms, len(pts)

    def state(self):
        return dict(a=self.a, b=self.b, lut_idx=self.lut_idx, calibrated=self.calibrated)


if __name__ == "__main__":
    r = Radiometry()
    print(f"loaded {r.n_lut} LUTs x {r.n_idx} entries")
    # self-consistency test: invent a ground-truth (lut, a, b), synthesize raw for known
    # temps, then check calibrate() recovers temperatures.
    gt = Radiometry(); gt.lut_idx = 3; gt.a = 35.0; gt.b = -250000.0
    true_temps = np.array([0.0, 20.0, 37.0, 60.0, 100.0])
    # invert model to get the raw that would yield each temp:
    target_rad = gt._radiance_at_U(gt.luts[gt.lut_idx], celsius_to_U(true_temps))
    raws = (target_rad - gt.b) / gt.a
    pts = list(zip(raws, true_temps))
    print("synthetic (raw, °C):", [(round(x, 1), t) for x, t in pts])
    rms, n = r.calibrate(pts)
    print(f"calibrated: lut={r.lut_idx} a={r.a:.4g} b={r.b:.4g} rms={rms:.4f}°C over {n} pts")
    print("recovered temps:", np.round(r.raw_to_celsius(raws), 3))
    # cross-check at an unseen raw
    test_raw = (gt._radiance_at_U(gt.luts[gt.lut_idx], celsius_to_U(np.array([50.0]))) - gt.b) / gt.a
    print("unseen 50°C -> predicted:", round(float(np.ravel(r.raw_to_celsius(test_raw))[0]), 3))

File: android/app/src/main/java/com/magnity/thermalcam/pipeline/Radiometry.kt
package com.magnity.thermalcam.pipeline

import com.magnity.thermalcam.data.NpyArray
import java.io.InputStream
import kotlin.math.floor
import kotlin.math.sqrt

/**
 * Temperature conversion — Kotlin port of `radiometry.py`, a faithful reimplementation
 * of the camera's built-in radiometric chain (reverse-engineered from libcoresdk.so):
 *
 *     v      = raw*(k+1) + b                       (folded here into rad = a*raw + b)
 *     target = v << (7 - shift)
 *     idx    = binary_search(radLUT, target)       radLUT = 646-entry Planck radiance LUT
 *     U      = idx*4096 - 150000 + (slopeLUT[idx]*(target-radLUT[idx])) >> 12
 *     °C     = U/1000 - 273.15                     (U is milli-Kelvin)
 *
 * The 8 built-in Planck curves come from `planck_luts.npy` (extracted from the APK's
 * .so); the per-camera linear term (a, b) is recovered from >=2 (raw, known °C)
 * reference points via [calibrate], exactly like the Linux app.
 */
class Radiometry(planckLutsNpy: InputStream) {

    companion object {
        const val U_OFFSET = -150000L
        const val U_STEP = 4096L
        const val KELVIN = 273.15
        const val U_PER_KELVIN = 1000.0

        fun celsiusToU(tc: Double): Double = (tc + KELVIN) * U_PER_KELVIN
        fun uToCelsius(u: Double): Double = u / U_PER_KELVIN - KELVIN
    }

    val nLut: Int
    val nIdx: Int
    private val luts: Array<LongArray>      // (8, 646) radiance curves
    private val slopes: Array<LongArray>    // firmware-exact: 2^24 / (L[i+1]-L[i])

    var a: Double = 1.0; private set
    var b: Double = 0.0; private set
    var lutIdx: Int = 0; private set
    var calibrated: Boolean = false; private set

    init {
        val arr = NpyArray.read(planckLutsNpy)
        nLut = arr.shape[0]
        nIdx = arr.shape[1]
        val flat = arr.toLongArray()
        luts = Array(nLut) { li -> LongArray(nIdx) { i -> flat[li * nIdx + i] } }
        slopes = Array(nLut) { li ->
            LongArray(nIdx - 1) { i ->
                val d = (luts[li][i + 1] - luts[li][i]).coerceAtLeast(1)
                (1L shl 24) / d
            }
        }
        lutIdx = nLut / 2
    }

    /** Restore a previously saved calibration. */
    fun restore(a: Double, b: Double, lutIdx: Int, calibrated: Boolean) {
        this.a = a; this.b = b
        this.lutIdx = lutIdx.coerceIn(0, nLut - 1)
        this.calibrated = calibrated
    }

    fun clear() { calibrated = false }

    // radiance at a given U (temperature-grid value), forward float interpolation
    private fun radianceAtU(lut: LongArray, u: Double): Double {
        var pos = (u - U_OFFSET) / U_STEP
        if (pos < 0) pos = 0.0
        if (pos > nIdx - 1.0001) pos = nIdx - 1.0001
        val i = pos.toInt()
        val frac = pos - i
        return lut[i] * (1 - frac) + lut[i + 1] * frac
    }

    /** Inverse radiance -> U, bit-faithful to the firmware's sub_3f970 simple path. */
    private fun uFromRadiance(li: Int, rad: Double): Double {
        val lut = luts[li]; val slope = slopes[li]
        var ti = floor(rad).toLong()
        if (ti < lut[0]) ti = lut[0]
        if (ti > lut[nIdx - 1]) ti = lut[nIdx - 1]
        // searchsorted(side='right') - 1, clamped to [0, nIdx-2]
        var lo = 0; var hi = nIdx        // first index with lut[idx] > ti
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            if (lut[mid] <= ti) lo = mid + 1 else hi = mid
        }
        val idx = (lo - 1).coerceIn(0, nIdx - 2)
        val delta = ti - lut[idx]
        val u = idx * U_STEP + U_OFFSET + ((slope[idx] * delta) shr 12)
        return u.toDouble()
    }

    /** raw counts -> °C via the calibrated linear term + the camera Planck LUT. */
    fun rawToCelsius(raw: Double): Double = uToCelsius(uFromRadiance(lutIdx, a * raw + b))

    /**
     * points: list of (raw, known °C). Tries every built-in LUT, least-squares fits
     * radiance = a*raw + b, keeps the LUT with the smallest °C residual.
     * Returns rms °C error.
     */
    fun calibrate(points: List<Pair<Double, Double>>): Double {
        require(points.size >= 2) { "need >= 2 reference points" }
        val raws = DoubleArray(points.size) { points[it].first }
        val temps = DoubleArray(points.size) { points[it].second }

        var bestRms = Double.MAX_VALUE
        var bestLi = 0; var bestA = 1.0; var bestB = 0.0
        for (li in 0 until nLut) {
            val lut = luts[li]
            val target = DoubleArray(raws.size) { radianceAtU(lut, celsiusToU(temps[it])) }
            // least-squares fit target = a*raw + b  (2x2 normal equations)
            val n = raws.size.toDouble()
            var sx = 0.0; var sy = 0.0; var sxx = 0.0; var sxy = 0.0
            for (i in raws.indices) {
                sx += raws[i]; sy += target[i]
                sxx += raws[i] * raws[i]; sxy += raws[i] * target[i]
            }
            val det = n * sxx - sx * sx
            if (det == 0.0) continue
            val fa = (n * sxy - sx * sy) / det
            val fb = (sy * sxx - sx * sxy) / det
            // residual measured back in °C
            var sse = 0.0
            for (i in raws.indices) {
                val pred = uToCelsius(uFromRadiance(li, fa * raws[i] + fb))
                sse += (pred - temps[i]) * (pred - temps[i])
            }
            val rms = sqrt(sse / raws.size)
            if (rms < bestRms) { bestRms = rms; bestLi = li; bestA = fa; bestB = fb }
        }
        a = bestA; b = bestB; lutIdx = bestLi; calibrated = true
        return bestRms
    }
}
